src/gluonnlp/attention_cell.py

```diff
@@ -0,0 +1,885 @@
+            Shape (query_length, batch_size, num_heads * value_units)
+    additional_info
+        scores:
+            Shape (batch_size, num_head, query_length, mem_length)
+        attn_weight:
+            Shape (batch_size, num_head, query_length, mem_length)
+    """
+    # TODO(sxjscience) Profile layout
+    if normalized:
+        query = l2_normalize(F, query, axis=-1, eps=eps)
+        key = l2_normalize(F, key, axis=-1, eps=eps)
+    if scaled:
+        if query_head_units is None:
+            query_shape = F.npx.shape_array(query)
+            # TODO(sxjscience) Remove .astype(np.float32).
+            #  Wait for https://github.com/apache/incubator-mxnet/issues/18084
+            scale = F.np.sqrt(query_shape[-1].astype(np.float32))
+        else:
+            scale = math.sqrt(query_head_units)
+        query = query / scale
+    if layout == 'NKT':
+        # 1. Expand the dimension of the mask:
+        #   (B, L_query, L_mem) --> (B, 1, L_query, L_mem)
+        if mask is not None:
+            mask = F.np.expand_dims(mask, axis=1)
+        # 2. Calculate the attention weights
+        #   Score: (B, N, L_query, C_Q) X (B, N, L_mem, C_Q) --> (B, N, L_query, L_mem)
+        scores = F.npx.batch_dot(query, key, transpose_b=True)
+        attn_weights = masked_softmax(F, scores, mask, axis=-1)
+        attn_weights = F.npx.dropout(attn_weights, p=dropout)
+        # 3. Calculate the context vector
+        # (B, N, L_query, L_mem) X (B, N, L_mem, C_V) --> (B, L_query, N * C_V)
+        if use_einsum:
+            context_vec = F.np.einsum('bnij,bnjc->binc', attn_weights, value)
+        else:
+            context_vec = F.npx.batch_dot(attn_weights, value).transpose((0, 2, 1, 3))
+        context_vec = F.npx.reshape(context_vec, (-2, -2, -1))
+    elif layout == 'NTK':
+        # 1. Expand the dimension of the mask:
+        #   (B, L_query, L_mem) --> (B, 1, L_query, L_mem)
+        if mask is not None:
+            mask = F.np.expand_dims(mask, axis=1)
+        # 2. Calculate the attention weights
+        #   Score: (B, L_query, N, C_Q) X (B, L_mem, N, C_Q) --> (B, N, L_query, L_mem)
+        if use_einsum:
+            scores = F.np.einsum('binc,bjnc->bnij', query, key)
+        else:
+            scores = F.npx.batch_dot(F.np.swapaxes(query, 1, 2), F.np.swapaxes(key, 1, 2),
+                                     transpose_b=True)
+        attn_weights = masked_softmax(F, scores, mask)
+        attn_weights = F.npx.dropout(attn_weights, p=dropout)
+        # 3. Calculate the context vector
+        # (B, N, L_query, L_mem) X (B, L_mem, N, C_V) --> (B, L_query, N * C_V)
+        if use_einsum:
+            context_vec = F.np.einsum('bnij,bjnc->binc', attn_weights, value)
+        else:
+            context_vec = F.npx.batch_dot(attn_weights,
+                                          F.np.swapaxes(value, 1, 2)).transpose((0, 2, 1, 3))
+        context_vec = F.npx.reshape(context_vec, (-2, -2, -1))
+    elif layout == 'TNK':
```

Comment:
@MoisesHer @ptrendx According to my understanding, the `contrib.interleaved_matmul_selfatt_qk` + `contrib.interleaved_matmul_selfatt_valatt`  or `contrib.interleaved_matmul_encdec_qk` + `interleaved_matmul_encdec_valatt` could be implemented as `einsum('ibnc,jbnc->bnij')` and `np.einsum('bnij,jbnc->ibnc')`